### PR TITLE
[Llama4] Use strided in-place fused QK RMSNorm to drop a redundant copy

### PR DESCRIPTION
**Pull request by kevin85421**
The qk_norm block reshape-then-contiguous() was forcing a copy because qk is a strided slice of the qkv buffer. Switch the CUDA path to apply_qk_norm, which calls the JIT fused_inplace_qknorm kernel — its TensorMatcher accepts arbitrary outer strides, so it reads/writes the qkv buffer in-place via the split q/k views and the reshape copy goes away. NPU/other backends keep the old path (qk is contiguous there after the torch.cat reconstruction, so the reshape is already a free view).

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

### CODE REVIEW

**Comment by kevin85421** [comment]
Not ready for review right now.

File: python/sglang/srt/models/llama4.py [hunk 1]
```diff
@@ -339,13 +340,27 @@ def forward(
...
             del q_view, k_view, q_out_unused, k_out_unused
 
-        if self.qk_norm is not None:
-            # TODO there are still 2 redundant direct_copy_kernel_cuda for this `reshape` and (in attn backend) q.contiguous(), maybe we can fuse them later
-            qk = qk.reshape(-1, self.head_dim).contiguous().bfloat16()
-            qk = self.qk_norm(qk).to(torch.bfloat16)
-            qk = qk.reshape(-1, self.q_size + self.kv_size)
-
-        q, k = qk.split([self.q_size, self.kv_size], dim=-1)
+        if self.qk_norm is not None and _is_cuda:
```
Comment (kevin85421) [inline python/sglang/srt/models/llama4.py:345]: To verify correctness, the test `test_strided_equivalence` ensures the outputs of `apply_qk_norm` and `self.qk_norm(qk)` are equivalent.


https://gist.github.com/kevin85421/c7d4958ed82341de7d5c7bd54af98372

Image

Comment (kevin85421) [inline python/sglang/srt/models/llama4.py:345]: Another verification is E2E tests:
* send 5 prompts to SGLang server with and without this PR.
  * Prompts
    1. "Q: What is 17 * 43?\nA:"
    2. "The capital of France is"
    3. "Once upon a time, in a faraway kingdom,"
    4. "def fibonacci(n):\n    if n < 2:\n        return n\n    return"
    5. "Explain the difference between a stack and a queue in two sentences."

The responses of (1) and (5) are exactly the same.

**Review by ch-wan — COMMENTED** [review]
Could you add some efficiency tests?

**Comment by kevin85421** [comment]
> Could you add some efficiency tests?

The benchmark results for 
(1) `bench_one_batch` and `bench_one_batch_server`
(2) with and without this PR

https://gist.github.com/kevin85421/65e201dc3ad33301d4146628572d0517

**Comment by kevin85421** [comment]
The CI failure appears to be related to the CI's auth setup — note the HTTP 403 in the screenshot.

image
